bindings/rust/s2n-tls-hyper/src/connector.rs

```diff
@@ -0,0 +1,234 @@
+        http.enforce_http(false);
+
+        Builder::new(http, conn_builder)
+    }
+}
+
+impl<T, B> HttpsConnector<T, B>
+where
+    B: connection::Builder,
+    <B as connection::Builder>::Output: Unpin,
+{
+    /// Creates a new `Builder` used to create an `HttpsConnector`.
+    ///
+    /// `conn_builder` will be used to produce the s2n-tls Connections used for negotiating HTTPS,
+    /// which can be an `s2n_tls::config::Config` or other `s2n_tls::connection::Builder`.
+    ///
+    /// This API allows a `Builder` to be constructed with an existing HTTP connector, as follows:
+    /// ```
+    /// use s2n_tls_hyper::connector::HttpsConnector;
+    /// use s2n_tls::config::Config;
+    /// use hyper_util::client::legacy::connect::HttpConnector;
+    ///
+    /// let mut http = HttpConnector::new();
+    ///
+    /// // Ensure that the HTTP connector permits the HTTPS scheme.
+    /// http.enforce_http(false);
+    ///
+    /// let builder = HttpsConnector::builder_with_http(http, Config::default());
+    /// ```
+    ///
+    /// `HttpsConnector::builder()` can be used to create a new HTTP connector automatically.
+    pub fn builder_with_http(http: T, conn_builder: B) -> Builder<T, B> {
+        Builder::new(http, conn_builder)
+    }
+}
+
+// hyper connectors MUST implement `hyper_util::client::legacy::connect::Connect`, which is an alias
+// for the `tower_service::Service` trait where `Service` is implemented for `http::uri::Uri`, and
+// `Service::Response` implements traits for compatibility with hyper:
+// https://docs.rs/hyper-util/latest/hyper_util/client/legacy/connect/trait.Connect.html
+//
+// The hyper compatibility traits for `Service::Response` are implemented in `MaybeHttpsStream`.
+impl<T, B> Service<Uri> for HttpsConnector<T, B>
+where
+    T: Service<Uri>,
+    T::Response: Read + Write + Connection + Unpin + Send + 'static,
+    T::Future: Send + 'static,
+    T::Error: Into<BoxError>,
+    B: connection::Builder + Send + Sync + 'static,
+    <B as connection::Builder>::Output: Unpin + Send,
+{
+    type Response = MaybeHttpsStream<T::Response, B>;
+    type Error = BoxError;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<MaybeHttpsStream<T::Response, B>, BoxError>> + Send>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        match self.http.poll_ready(cx) {
+            Poll::Ready(Ok(())) => Poll::Ready(Ok(())),
+            Poll::Ready(Err(e)) => Poll::Ready(Err(e.into())),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+
+    fn call(&mut self, req: Uri) -> Self::Future {
+        // Currently, the only supported stream type is TLS. If the application attempts to
+        // negotiate HTTP over plain TCP, return an error.
+        if req.scheme() == Some(&http::uri::Scheme::HTTP) {
+            return Box::pin(async move { Err(UnsupportedScheme.into()) });
+        }
```

Comment:
HTTP support is needed so that users can create a single hyper client that will choose the right protocol based on the provided scheme in the URL. Otherwise, users that want to send a request over HTTP and a request over HTTPS would have to create two separate hyper clients. This would be a breaking change for users switching from another hyper connector crate to s2n-tls-hyper.

> And in the HTTP case could we just return whatever stream currently implements plain HTTP instead of needing the MaybeHttpsStream? Aren't we only interested in special behavior for HTTPS?

My understanding is that this is sort of what the MaybeHttpsStream is for - returning the plain HTTP stream instead of the s2n-tls stream for the HTTP case. The issue is that `tower_service::Service` requires the response type to be declared as a associated type: https://github.com/aws/s2n-tls/blob/e9a5c4d5300d74fd117144862453acfb3fb5915e/bindings/rust/s2n-tls-hyper/src/connector.rs#L128

So in order to return two different types of streams, they need to be wrapped together into a single type, which the MaybeHttpsStream will do.